nifi-nar-bundles/nifi-hive-bundle/nifi-hive3-processors/src/main/java/org/apache/hive/streaming/NiFiRecordSerDe.java

```diff
@@ -0,0 +1,247 @@
+
+    public NiFiRecordSerDe(RecordReader recordReader, ComponentLog log) {
+        this.recordReader = recordReader;
+        this.log = log;
+    }
+
+    @Override
+    public void initialize(Configuration conf, Properties tbl) {
+        List<TypeInfo> columnTypes;
+        StructTypeInfo rowTypeInfo;
+
+        log.debug("Initializing NiFiRecordSerDe: {}", tbl.entrySet().toArray());
+
+        // Get column names and types
+        String columnNameProperty = tbl.getProperty(serdeConstants.LIST_COLUMNS);
+        String columnTypeProperty = tbl.getProperty(serdeConstants.LIST_COLUMN_TYPES);
+        final String columnNameDelimiter = tbl.containsKey(serdeConstants.COLUMN_NAME_DELIMITER) ? tbl
+                .getProperty(serdeConstants.COLUMN_NAME_DELIMITER) : String.valueOf(SerDeUtils.COMMA);
+        // all table column names
+        if (columnNameProperty.isEmpty()) {
+            columnNames = new ArrayList<>(0);
+        } else {
+            columnNames = new ArrayList<>(Arrays.asList(columnNameProperty.split(columnNameDelimiter)));
+        }
+
+        // all column types
+        if (columnTypeProperty.isEmpty()) {
+            columnTypes = new ArrayList<>(0);
+        } else {
+            columnTypes = TypeInfoUtils.getTypeInfosFromTypeString(columnTypeProperty);
+        }
+
+        log.debug("columns: {}, {}", new Object[]{columnNameProperty, columnNames});
+        log.debug("types: {}, {} ", new Object[]{columnTypeProperty, columnTypes});
+
+        assert (columnNames.size() == columnTypes.size());
+
+        rowTypeInfo = (StructTypeInfo) TypeInfoFactory.getStructTypeInfo(columnNames, columnTypes);
+        schema = rowTypeInfo;
+        log.debug("schema : {}", new Object[]{schema});
+        cachedObjectInspector = (StandardStructObjectInspector) TypeInfoUtils.getStandardJavaObjectInspectorFromTypeInfo(rowTypeInfo);
+        tsParser = new TimestampParser(HiveStringUtils.splitAndUnEscape(tbl.getProperty(serdeConstants.TIMESTAMP_FORMATS)));
+    }
+
+    @Override
+    public Class<? extends Writable> getSerializedClass() {
+        return null;
+    }
+
+    @Override
+    public Writable serialize(Object o, ObjectInspector objectInspector) throws SerDeException {
+        return null;
+    }
+
+    @Override
+    public SerDeStats getSerDeStats() {
+        return null;
+    }
+
+    @Override
+    public Object deserialize(Writable writable) throws SerDeException {
+        ObjectWritable t = (ObjectWritable) writable;
+        Record record = (Record) t.get();
+        List<Object> r = new ArrayList<>(Collections.nCopies(columnNames.size(), null));
+        try {
+            RecordSchema recordSchema = record.getSchema();
+            for (RecordField field : recordSchema.getFields()) {
+                String fieldName = field.getFieldName();
+                int fpos = schema.getAllStructFieldNames().indexOf(fieldName.toLowerCase());
```

Comment:
Once the processor is started can the record schema change for different records?
If not can this can field name to position mapping be moved to init?

Reply:
No the schema will be the same for all records, I thought about moving it to init but wasn't sure about the Hive internal columns, I can certainly try moving it to init() (unless we'll remove this entirely based on your next comment)

Reply:
I don't understand the part about the internal columns. The destination table will never have internal column names. What would be the case for special casing the internal columns?

Reply:
I took this code from the StrictJsonWriter, it tries to find internal column names from the JSON fields. I wasn't sure what it was used for, so I kept it in so as to make the behavior consistent. Happy to remove it if we don't need it.